implementations/micrometer-registry-new-relic/src/main/java/io/micrometer/newrelic/NewRelicHttpClientProviderImpl.java

```diff
@@ -0,0 +1,296 @@
+                )
+            );
+    }
+
+    @Override
+    public Stream<String> writeMeter(Meter meter) {
+        // Snapshot values should be used throughout this method as there are chances for values to be changed in-between.
+        Map<String, Attribute> attributes = new HashMap<>();
+        for (Measurement measurement : meter.measure()) {
+            double value = measurement.getValue();
+            if (!Double.isFinite(value)) {
+                continue;
+            }
+            String name = measurement.getStatistic().getTagValueRepresentation();
+            attributes.put(name, new Attribute(name, value));
+        }
+        if (attributes.isEmpty()) {
+            return Stream.empty();
+        }
+        return Stream.of(event(meter.getId(), attributes.values().toArray(new Attribute[0])));
+    }
+
+    private String event(Meter.Id id, Attribute... attributes) {
+        if (!config.meterNameEventTypeEnabled()) {
+            // Include contextual attributes when publishing all metrics under a single categorical eventType,
+            // NOT when publishing an eventType per Meter/metric name
+            int size = attributes.length;
+            Attribute[] newAttrs = Arrays.copyOf(attributes, size + 2);
+
+            String name = id.getConventionName(namingConvention);
+            newAttrs[size] = new Attribute(METRIC_NAME, name);
+            newAttrs[size + 1] = new Attribute(METRIC_TYPE, id.getType().toString());
+            
+            return event(id, Tags.empty(), newAttrs);
+        }
+        return event(id, Tags.empty(), attributes);
+    }
+
+    private String event(Meter.Id id, Iterable<Tag> extraTags, Attribute... attributes) {
+        StringBuilder tagsJson = new StringBuilder();
+
+        for (Tag tag : id.getConventionTags(namingConvention)) {
+            tagsJson.append(",\"").append(escapeJson(tag.getKey())).append("\":\"").append(escapeJson(tag.getValue())).append("\"");
+        }
+
+        for (Tag tag : extraTags) {
+            tagsJson.append(",\"").append(escapeJson(namingConvention.tagKey(tag.getKey())))
+                    .append("\":\"").append(escapeJson(namingConvention.tagValue(tag.getValue()))).append("\"");
+        }
+
+        String eventType = getEventType(id, config, namingConvention);
+        
+        return Arrays.stream(attributes)
+                .map(attr -> 
+                        (attr.getValue() instanceof Number) 
+                            ? ",\"" + attr.getName() + "\":" + DoubleFormat.wholeOrDecimal(((Number)attr.getValue()).doubleValue())
+                            : ",\"" + attr.getName() + "\":\"" + namingConvention.tagValue(attr.getValue().toString()) + "\""
+                )
+                .collect(Collectors.joining("", "{\"eventType\":\"" + escapeJson(eventType) + "\"", tagsJson + "}"));
+    }
+
+    void sendEvents(Meter.Id id, Object attributesObj) {
```

Comment:
This looks like it could be simplified by changing the type of the second parameter to `Stream<String>` instead of `Object`.

Reply:
Updated.